[OpenMPIRBuilder] Don't drop debug info for target region.

abidh opened the pull request:
When an outlined function is generated for omp target region, a corresponding DISubprogram was not being generated. This resulted in all the debug information for the target region being dropped.

This commit adds DISubprogram for the outlined function if there is one available for the parent function. It also updates the current debug location so that the right scope is used for the entries in the outlined function.

With this change in place, I can set source line breakpoint in target region and run to them in debugger.

abidh committed:
[OpenMPIRBuilder] Don't drop debug info for target region.

When an outlined function is generated for omp target region,
a corresponding DISubprogram was not being generated. This resulted in
all the debug information for the target region being dropped.

This commit adds DISubprogram for the outlined function if there is one
available for the parent function. It also updates the current debug
location so that the right scope is used for the entries in the outlined
function.

With this change in place, I can set source line breakpoint in target
region and run to them in debugger.

DominikAdamski commented on flang/test/Lower/OpenMP/target-debug.f90:
Is it possible to generate MLIR file and add test for LLVM IR [here](https://github.com/llvm/llvm-project/blob/main/mlir/test/Target/LLVMIR/) ?

abidh replied:
Done.

DominikAdamski reviewed (commented):
Is it possible to generate MLIR file and

jdoerfert commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5025,10 +5026,41 @@ static Function *createOutlinedFunction(
+
+  // If there's a DISubprogram associated with current function, then
+  // generate one for the outlined function.
+  if (Function *parentFunc = BB->getParent()) {
```
Do we expect this to be an unconnected BB at any point?
Here, and below, please use LLVM style. This is not MLIR.

abidh replied:
I don't think BB will be unconnected. But even if that was to happen, the worst we will do is to drop debug bits as we are already doing. 

Thanks for pointing out the style bit. Corrected now.

abidh committed:
Move the test into MLIR.

The original f90 ->llvmir test has been removed.
 I have added an MLIR -> llvmir test as per review comments.

abidh committed:
Correct the case on variable name.

abidh commented:
Ping. Any further review comments for this PR?

abidh commented:
Ping.

mjklemm commented:
@jdoerfert @DominikAdamski @kiranchandramohan Kindly requesting to review this PR, so that it can land.

kiranchandramohan commented on mlir/test/Target/LLVMIR/omptarget-debug.mlir:
```diff
@@ -0,0 +1,72 @@
+module attributes {omp.is_target_device = true} {
+  llvm.func @_QQmain() {
+    %0 = llvm.mlir.constant(1 : i64) : i64
+    %1 = llvm.alloca %0 x i32 : (i64) -> !llvm.ptr<5>
+    %2 = llvm.addrspacecast %1 : !llvm.ptr<5> to !llvm.ptr
+    %3 = llvm.mlir.constant(1 : i64) : i64
+    %4 = llvm.alloca %3 x i32 : (i64) -> !llvm.ptr<5>
+    %5 = llvm.addrspacecast %4 : !llvm.ptr<5> to !llvm.ptr
+    %6 = llvm.mlir.constant(1 : i64) : i64
+    %7 = llvm.alloca %6 x i32 : (i64) -> !llvm.ptr<5>
+    %8 = llvm.addrspacecast %7 : !llvm.ptr<5> to !llvm.ptr
+    %9 = omp.map_info var_ptr(%2 : !llvm.ptr, i32) map_clauses(tofrom) capture(ByRef) -> !llvm.ptr {name = ""}
+    %10 = omp.map_info var_ptr(%5 : !llvm.ptr, i32) map_clauses(tofrom) capture(ByRef) -> !llvm.ptr {name = ""}
+    omp.target map_entries(%9 -> %arg0, %10 -> %arg1 : !llvm.ptr, !llvm.ptr) {
+    ^bb0(%arg0: !llvm.ptr, %arg1: !llvm.ptr):
+      %12 = llvm.mlir.constant(2 : i32) : i32
+      %13 = llvm.mlir.constant(1 : i32) : i32
+      %14 = llvm.load %arg0 : !llvm.ptr -> i32 loc(#loc2)
+      %15 = llvm.add %14, %13  : i32 loc(#loc2)
+      llvm.store %15, %arg0 : i32, !llvm.ptr loc(#loc2)
+      %16 = llvm.load %arg0 : !llvm.ptr -> i32 loc(#loc3)
+      %17 = llvm.add %16, %12  : i32 loc(#loc3)
+      llvm.store %17, %arg1 : i32, !llvm.ptr loc(#loc3)
+      omp.terminator
+    }
+    %11 = omp.map_info var_ptr(%8 : !llvm.ptr, i32) map_clauses(tofrom) capture(ByRef) -> !llvm.ptr {name = ""}
+    omp.target map_entries(%11 -> %arg0 : !llvm.ptr) {
+    ^bb0(%arg0: !llvm.ptr):
+      %12 = llvm.mlir.constant(1 : i32) : i32
+      omp.parallel {
+        %13 = llvm.load %arg0 : !llvm.ptr -> i32 loc(#loc4)
+        %14 = llvm.add %13, %12  : i32 loc(#loc4)
+        llvm.store %14, %arg0 : i32, !llvm.ptr loc(#loc4)
+        omp.terminator
+      }
+      omp.terminator
+    }
+    llvm.return
+  } loc(#loc5)
+} loc(#loc)
```
Nit: can you minimize this testcase to be as small as possible? We do not need a large region particularly.

kiranchandramohan reviewed (approved):
LG.

abidh committed:
Minimize testcase as per review comments.

abidh committed:
Don't generate debug info for variable in target region.

These variables currently fail the llvm verifier tests as the scope
inside the variable is still point to the parent function of the
target region.

abidh committed:
Improve test case.

abidh committed:
Use InsertPointGuard instead of manual save/restore.

It has the extra benefit of saving/restoring the debug location which otherwise could be lost.

abidh committed:
Check DebugLoc is valid before using it.

abidh committed:
Add testcase with is_target_device = false

abidh commented:
I have improved the patch a bit more and added some more tests especially to check that variable in target region don't cause a problem. I have tested it further and it seems ready to merge. Please let me know if there are any further comments.

abidh commented:
@jeanPerier Can you kindly look at the this PR especially the flang changes where I try to skip the debug info generation for variables which will be in OpenMP target region. Is this the right way to do it? I thought about processing DeclOp just in the function entry block which will have the same effect but this seems to make the intention more explicit.

abidh committed:
Avoid using stale debug location.
